src/components/power-sensor/client.ts

```diff
@@ -0,0 +1,110 @@
+import { Struct } from 'google-protobuf/google/protobuf/struct_pb';
+import type { RobotClient } from '../../robot';
+import { PowerSensorServiceClient } from '../../gen/component/powersensor/v1/powersensor_pb_service';
+import type { Options, StructType } from '../../types';
+import pb from '../../gen/component/powersensor/v1/powersensor_pb';
+import { promisify, doCommandFromClient } from '../../utils';
+import type { PowerSensor, PowerSensorReadings } from './power-sensor';
+
+/**
+ * A gRPC-web client for the MovementSensor component.
+ *
+ * @group Clients
+ */
+
+export class PowerSensorClient implements PowerSensor {
+  private client: PowerSensorServiceClient;
+  private readonly name: string;
+  private readonly options: Options;
+
+  constructor(client: RobotClient, name: string, options: Options = {}) {
+    this.client = client.createServiceClient(PowerSensorServiceClient);
+    this.name = name;
+    this.options = options;
+  }
+
+  private get powersensorService() {
+    return this.client;
+  }
+
+  async getVoltage(extra = {}) {
+    const { powersensorService } = this;
+    const request = new pb.GetVoltageRequest();
+    request.setName(this.name);
+    request.setExtra(Struct.fromJavaScript(extra));
+
+    this.options.requestLogger?.(request);
+
+    const response = await promisify<
+      pb.GetVoltageRequest,
+      pb.GetVoltageResponse
+    >(powersensorService.getVoltage.bind(powersensorService), request);
+
+    return [response.getVolts(), response.getIsAc()] as const;
+  }
+
+  async getCurrent(extra = {}) {
+    const { powersensorService } = this;
+    const request = new pb.GetCurrentRequest();
+    request.setName(this.name);
+    request.setExtra(Struct.fromJavaScript(extra));
+
+    this.options.requestLogger?.(request);
+
+    const response = await promisify<
+      pb.GetCurrentRequest,
+      pb.GetCurrentResponse
+    >(powersensorService.getCurrent.bind(powersensorService), request);
+
+    return [response.getAmperes(), response.getIsAc()] as const;
+  }
+
+  async getPower(extra = {}) {
+    const { powersensorService } = this;
+    const request = new pb.GetPowerRequest();
+    request.setName(this.name);
+    request.setExtra(Struct.fromJavaScript(extra));
+
+    this.options.requestLogger?.(request);
+
+    const response = await promisify<pb.GetPowerRequest, pb.GetPowerResponse>(
+      powersensorService.getPower.bind(powersensorService),
+      request
+    );
+    return response.getWatts();
+  }
+
+  async getReadings(extra = {}) {
+    const readingFunctions = {
+      voltage: this.getVoltage.bind(this),
+      current: this.getCurrent.bind(this),
+      power: this.getPower.bind(this),
+    };
```

Comment:
assuming voltage and current will always have the current flow type for a given power source, maybe the following is a better return type:

```ts
{
  isAc: boolean,
  voltage: number,
  current: number,
  power: number,
}
```

(and we can throw an error if the two `isAc` values ever differ).

Reply:
yes they should always be same, in go we decided to just return one of the isAC values and not check if they are different since that would be a driver error.  I'll change to do the same in TS